x-pack/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -243,12 +244,14 @@ const getWaterfallItems = (items: TraceAPIResponse['trace']['items']) =>
  * @param waterfallItems
  */
 const reparentSpans = (waterfallItems: IWaterfallItem[]) => {
```

Comment:
Instead of cloning you could make the function pure by splitting it into two operations:
1.  find children that needs to be updated
2. iterate over waterfallItems and update as needed

While a little longer I find this easier to comprehend - but I'm find with what you already have if you prefer that. Just food for thought :)

```ts
const reparentSpans = (waterfallItems: IWaterfallItem[]): IWaterfallItem[] => {
  // find children that needs to be re-parented and map them to their correct parent id
  const childIdToParentIdMapping = flatten(
    waterfallItems.map(waterfallItem => {
      if (waterfallItem.docType === 'span') {
        const childIds = waterfallItem.doc.child?.id ?? [];
        return childIds.map(id => ({ [id]: waterfallItem.id }));
      }
      return [];
    })
  );

  // update parent id for children that needs it or return unchanged
  return waterfallItems.map(waterfallItem => {
    const newParentId = childIdToParentIdMapping[waterfallItem.id];
    if (newParentId) {
      return {
        ...waterfallItem,
        parentId: newParentId
      };
    }

    return waterfallItem;
  });
};
```

Btw untested so might be a typo here or there

Reply:
changed here -> https://github.com/elastic/kibana/pull/65840/commits/838c0acc617e11b3ad2ee2617d374acb48a791ac 

Thanks for the feedback.